### PR TITLE
feat: Allow for the deletion of any document

### LINKED ISSUE #696: [DOC-372] Allow for the deletion of any document
### **Description**

Expand deletion functionality to include documents in PENDING and COMPLETED states, alongside existing draft deletion.

### **Improvement Description**

* Allow deletion of documents in any state (DRAFT, PENDING, COMPLETED).
* Implement varying confirmation requirements based on the document state.
* Soft delete documents using a `deletedAt` column.

### **Rationale**

We've had a couple of users complain that they are unable to delete an erroneous document once it's left the draft state, this has hindered their ability to use Documenso and doesn't match that of other providers. To improve user satisfaction and match our competition, we should improve our current handling of deletions. It also hinders the user from handling their data as they please.

### **Proposed Solution**

* Maintain simple deletion for drafts.
* Require explicit confirmation (e.g., typing "delete me") for deleting PENDING or COMPLETED documents.
* Email recipients for documents that are PENDING so they are aware that it has been canceled.
* Implement soft deletion for documents, retaining access in the inbox.

### **Alternatives**

* Introduce a temporary holding area for deleted documents before permanent deletion. (Recycle Bin)
* User-customizable settings for deletion confirmations.

<sub>From [SyncLinear.com](https://synclinear.com) | [DOC-372](https://linear.app/documenso/issue/DOC-372/allow-for-the-deletion-of-any-document)</sub>

### PR DESCRIPTION
**Pull request by ksushant6566**
Resolves #696

### CODE REVIEW

File: packages/trpc/server/document-router/router.ts
Comment (Mythie) [inline packages/trpc/server/document-router/router.ts:113]: Should we simplify this by merging the two and just returning early for draft documents? We typically don't introduce logic in our tRPC routers as doing so makes them less portable for other API types.

Comment (ksushant6566) [reply packages/trpc/server/document-router/router.ts:113]: oh okay, I'll just make a `deleteDocument({ id, userId, status })` function that merges the functionality

File: apps/web/src/app/(dashboard)/documents/deleteWithConfirmation.tsx
Comment (Mythie) [inline apps/web/src/app/(dashboard)/documents/deleteWithConfirmation.tsx:3]: This file doesn't appear to be used? 🤔

File: packages/email/templates/document-cancel.tsx
Comment (Mythie) [inline packages/email/templates/document-cancel.tsx:16]: Feels like a redundant template string since we don't change it based on any conditions? 🤔

File: packages/lib/server-only/document/delete-document.ts
Comment (Mythie) [inline packages/lib/server-only/document/delete-document.ts:19]: Good stuff on making this change 😄

**Comment by Mythie** [comment]
I think this is mostly missing interactions with what happens when a signer visits the signing page for a deleted document etc.

If a document was deleted do we redirect them? Show an error page? If we do either of those what do we do for completed documents with a valid token, should we just show the normal completion page?

File: apps/web/src/app/(dashboard)/documents/delete-document-dialog.tsx
Comment (ksushant6566) [reply apps/web/src/app/(dashboard)/documents/delete-document-dialog.tsx:55]: yeah you are right!, lemme just make these changes and push quickly

Comment (ksushant6566) [reply apps/web/src/app/(dashboard)/documents/delete-document-dialog.tsx:55]: wait we are toasting on successful deletes, see OnSuccess of `trpcReact.document.deleteDocument.useMutation`

File: apps/web/src/app/(dashboard)/documents/delete-document-dialog.tsx
Comment (ksushant6566) [reply apps/web/src/app/(dashboard)/documents/delete-document-dialog.tsx:66]: okay, changing this

File: apps/web/src/app/(dashboard)/documents/deleteWithConfirmation.tsx
Comment (ksushant6566) [reply apps/web/src/app/(dashboard)/documents/deleteWithConfirmation.tsx:3]: oops

File: packages/email/templates/document-cancel.tsx
Comment (ksushant6566) [reply packages/email/templates/document-cancel.tsx:16]: In my opinion, it just makes it more readable & future-proof. But lmk if you want it changed?

File: packages/lib/server-only/document/delete-document.ts
Comment (ksushant6566) [reply packages/lib/server-only/document/delete-document.ts:41]: I mean we can do that, but maybe this seems a bit more readable?

File: packages/lib/server-only/document/delete-document.ts
Comment (ksushant6566) [reply packages/lib/server-only/document/delete-document.ts:48]: oh yes, if there are no recipients then we don't need to do anything

**Comment by ksushant6566** [comment]
> I think this is mostly missing interactions with what happens when a signer visits the signing page for a deleted document etc.
> 
> If a document was deleted do we redirect them? Show an error page? If we do either of those what do we do for completed documents with a valid token, should we just show the normal completion page?

I think for completed documents we can just show the normal page & for pending documents we can disable actions and show a message that **_this document is no longer available to sign_**

wdyt?

File: packages/lib/server-only/document/delete-document.ts
Comment (Mythie) [reply packages/lib/server-only/document/delete-document.ts]: That's what I was thinking :)

File: packages/lib/server-only/document/delete-document.ts
Comment (Mythie) [reply packages/lib/server-only/document/delete-document.ts]: This will involve two round trips to the database while there's potential for the one above to do a join, it's negligible in terms of performance but could add up later on. Happy to leave for now for the sake of readability though 😄

File: packages/email/templates/document-cancel.tsx
Comment (Mythie) [reply packages/email/templates/document-cancel.tsx:16]: Happy to keep, just noting that it doesn't do much 😄 

I try to avoid designing for a possible future and instead design for the today and the potential throw away

**Comment by Mythie** [comment]
I'll do a round of UI review with this now as well, I've only been stinging you on code so far 👀

**Comment by Mythie** [comment]
This will also need some guards for recipients adding signatures when a document has since been deleted 😄

**Comment by ksushant6566** [comment]
> This will also need some guards for recipients adding signatures when a document has since been deleted 😄

Done, added a check in `sign-field-with-token.ts` .

**Comment by ksushant6566** [comment]
Here is a loom video of the solution :)

https://www.loom.com/share/8f28b9ad4bb64a70b33da9d59fd2a95b?sid=3da77147-be41-486a-af86-cc8fd09afce2

File: apps/web/src/app/(dashboard)/documents/delete-document-dialog.tsx
Comment (Mythie) [reply apps/web/src/app/(dashboard)/documents/delete-document-dialog.tsx:55]: Right I always get mixed up since we've changed things up a little there!

File: apps/web/src/app/(dashboard)/documents/delete-document-dialog.tsx
Comment (Mythie) [inline apps/web/src/app/(dashboard)/documents/delete-document-dialog.tsx:106]: This should be a `variant="destructive"` so we know we're nuking something

File: apps/web/src/app/(signing)/sign/[token]/complete/page.tsx
Comment (Mythie) [inline apps/web/src/app/(signing)/sign/[token]/complete/page.tsx:102]: Same thing

**Comment by ksushant6566** [comment]
Yo, I believe everything works now, but still do confirm if the logic below makes sense:

What documents users can see:

**own documents**
_All documents created by me, but not the ones I deleted._

**received documents**

Pending documents: 

- not deleted: Yes
- deleted but signed: Yes
- deleted, not signed (Inbox):  No

Completed documents: 

- not deleted: Yes
- deleted: Yes

Note: Only owners can delete documents, users can't delete received documents. That will involve lot more work. (Maybe open another issue?)

PS: thanks a lot for your help throughout 🙌

**Comment by Mythie** [comment]
This all sounds correct to me, going to run through a few rounds of testing before I hit the merge button :)

**Review by Mythie — APPROVED** [review]

**Pull request merged**
